Warn user if client certificate is expired

cmurphy opened the pull request:
If a user's client certificate is expired, the API server will reject a
kubectl request with a vague 401 error. Adding more information to the
server's response risks exposing too much information to a potential
attacker, but it's reasonable to add client-side validation prior making
the request. This change to client-go parses the user's certificate upon
loading it and checks whether it is valid according to the host's system
time. If the certificate is expired, it logs a warning but allows the
request to continue; this is to accomodate the case that the user's host
system may not have the correct time set but the certificate is still
valid to the API server, while still leaving an appropriate hint to the
user.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

/sig auth
/sig cli

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #39767

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added client-side validation of client certificate. Kubectl will now warn users if their client certificate is expired.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

cmurphy committed:
Warn user if client certificate is expired

If a user's client certificate is expired, the API server will reject a
kubectl request with a vague 401 error. Adding more information to the
server's response risks exposing too much information to a potential
attacker, but it's reasonable to add client-side validation prior making
the request. This change to client-go parses the user's certificate upon
loading it and checks whether it is valid according to the host's system
time. If the certificate is expired, it logs a warning but allows the
request to continue; this is to accomodate the case that the user's host
system may not have the correct time set but the certificate is still
valid to the API server, while still leaving an appropriate hint to the
user.

jktomer commented:
/remove-sig api-machinery

tanjunchen commented:
/ok-to-test

cmurphy commented:
/retest

cmurphy commented:
Added a new commit to simplify the warning function. Will need to be squashed.

cmurphy commented:
/remove-sig api-machinery

cmurphy commented:
/retest

enj commented on staging/src/k8s.io/client-go/transport/transport.go:
```diff
@@ -119,6 +124,24 @@ func TLSConfigFor(c *Config) (*tls.Config, error) {
+		parsedCert, err := x509.ParseCertificate(certData)
+		if err != nil {
+			return err
+		}
```
I do not think this should error out at all:

`func warnIfExpired(c *tls.Certificate) {`

and just log any `ParseCertificate` error and continue.

boddumanohar replied:
@enj this could error out if there is a issue with parsing right?

enj replied:
> @enj this could error out if there is a issue with parsing right?

No, I think it should just warn along the lines of "I see you are using a client cert but I cannot seem to parse it to determine if it has expired."

I do not want a change in control flow.

enj reviewed (commented):
I still need to figure out if this is the best spot for logging such an error (seems like it should be in `kubectl` specific code and not generic transport code in `client-go`).

I think the overall idea of warning the user for this is not a bad idea, but would like opinions from others.

It should also be possible to write a BASH test with `kubectl` to show that the warnings get logged when the cert is expired and vice versa.

soltysh commented:
> I still need to figure out if this is the best spot for logging such an error (seems like it should be in `kubectl` specific code and not generic transport code in `client-go`).

Since `kubectl` was called out I'll add my 2 cents, theoretically all of config loading is happening right here in client-go's code, like @cmurphy nicely pointed out and attempted a fix. `kubectl` is using the code that lives in https://github.com/kubernetes/cli-runtime/blob/master/pkg/genericclioptions/config_flags.go to go from flags passed by user to an actual config. Maybe that'd be a good place for this?

cmurphy commented:
@enj - thanks for the suggestion, I added a bash test for this.

@soltysh - this latest commit follows your suggestion and moves the logic to the genericclioptions package in cli-runtime. However, I think it is somewhat problematic. Most of the configuration loading and validation currently happens in client-go, especially https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/validation.go. If I try to load the certificate in cli-runtime before that validation has happened, invalid data in the kubeconfig won't be handled gracefully and will instead just cause kubectl to segfault. Additionally, to make this work I've had to copy a lot of logic that currently already exists in client-go's transport package, so both the code and the processing work is being duplicated. Is this what you had in mind, and do you have a suggestion on how I could improve on this?

cmurphy commented:
> @soltysh - this latest commit follows your suggestion and moves the logic to the genericclioptions package in cli-runtime. However, I think it is somewhat problematic. Most of the configuration loading and validation currently happens in client-go, especially https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/validation.go. If I try to load the certificate in cli-runtime before that validation has happened, invalid data in the kubeconfig won't be handled gracefully and will instead just cause kubectl to segfault. Additionally, to make this work I've had to copy a lot of logic that currently already exists in client-go's transport package, so both the code and the processing work is being duplicated. Is this what you had in mind, and do you have a suggestion on how I could improve on this?

It also seems like the kubectl proxy integration tests are failing because of klog trying to use stdout here, whereas there was no issue with client-go using klog.

cmurphy commented:
/retest

cmurphy commented:
@soltysh with the code freeze lifted, would you mind giving this another look and giving some thought to my concerns in https://github.com/kubernetes/kubernetes/pull/87796#issuecomment-593582851 ?

cmurphy commented:
Rebased to resolve merge conflicts. I would really appreciate reviews, especially with regard to my concern noted in https://github.com/kubernetes/kubernetes/pull/87796#issuecomment-593582851. I am committed to seeing this through. Not sure if it's appropriate to `/assign` an approver because I do not think this is ready to be approved, as I've noted.

cmurphy commented:
/retest

cmurphy commented:
/retest

cmurphy commented:
/retest

cmurphy commented:
It is difficult to keep rebasing a long-ish series of fixup commits so I have squashed them. A copy of the old branch before this latest update is [here](https://github.com/cmurphy/kubernetes/tree/expired-cert-warnings-39767-draft). Since there was no feedback since I moved the logic to cli-runtime I reverted back to putting it in client-go because the alternative seemed really clunky (https://github.com/kubernetes/kubernetes/pull/87796#issuecomment-593582851).

cmurphy commented:
/remove-sig api-machinery

cmurphy commented:
/retest

cmurphy commented:
/remove-sig api-machinery
/remove-lifecycle rotten

cmurphy commented:
/retest
